File: cpp/456_132模式_medium.cpp
class Solution {
public:
    bool find132pattern(vector<int>& nums) {
        int N = nums.size();
        if (N < 3)  return false;
        stack<int> candidate_k;
        candidate_k.push(nums[N-1]);
        int max_k = INT_MIN;
        for(int i=N-2; i>=0; i--) {
            if (nums[i] < max_k)
                return true;
            while (!candidate_k.empty() && nums[i]>candidate_k.top()) {
                max_k = candidate_k.top();
                candidate_k.pop();
            }
            if (nums[i]>max_k)
                candidate_k.push(nums[i]);
        }
        return false;
    }
};

File: python/456_132模式_medium.py
class Solution:
    def find132pattern(self, nums: List[int]) -> bool:
        N = len(nums)
        candidate_k = [nums[N-1]]
        max_k = float("-inf")
        for i in range(N-2, -1, -1):
            if nums[i] < max_k:
                return True
            while candidate_k and nums[i] > candidate_k[-1]:
                max_k = candidate_k[-1]
                candidate_k.pop()
            if nums[i] > max_k:
                candidate_k.append(nums[i])
        return False
